### DISCUSSION TITLE
Can I use Sparkle to update my JVM application?

### DISCUSSION THREAD

**Discussion by ScottPierce** [sparkle-project/Sparkle#2239] [Q&A]
I'm releasing a JVM application that doesn't live inside of an xcode project. It compiles via jpackage to a .app folder that is distributed via DMG.

How could I integrate Sparkle with my app to update my app, that doesn't live inside an xcode project?

**Comment by zorgiepoo** [comment]
Sparkle doesn't require Xcode-specific things. You'd have to look at the [programmatic setup](https://sparkle-project.org/documentation/programmatic-setup/) and updater APIs available in addition to the rest of the documentation.

This means you will have to figure out how to bundle another Obj-C framework in your app and talk to Obj-C APIs. At worst you may have to go through JNI. Edit: you may also need to ensure the runloop is set up correctly..

I imagine you'd similarly already have to be going through some hoops like bundling the JRE in your app so users don't have to install Java, which is what modern Java based applications do these days on the Mac.

This is the general approach anyway that other UI toolkits (e.g. Qt also) do to use Sparkle anyway. I believe one of Oracle's apps use Sparkle and apps like Cyberduck use Sparkle (although it has a cocoa frontend).

**Reply by ScottPierce** [reply]
@zorgiepoo ObjC + JNI is very complex because ObjC requires additional setup from a normal application I believe.

What if I created a separate ObjC application with the sparkle framework installed? Then I could call into it to perform an update? Would it make sense to do this?

**Reply by zorgiepoo** [reply]
The "hurdle" here is Java calling out to native code (Java is the "not-normal" bit here). You need to figure out how to load a dynamic library, export and call a C function at minimum. [Here is a gist](https://gist.github.com/zorgiepoo/6f81e0d902c00722ce992029f3f82589) of building a custom dylib that links to Sparkle and starts the updater with a minimal setup. The missing bit is dlopen'ing the library or whatever equivalent it is in Java through JNI.

You can technically run the updater in a separate app/process that targets another bundle but this should be a last resort after you make a decent attempt to not have to do this. It may be harder to get the updater to "feel right" this way.
